Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py

```diff
@@ -139,14 +153,65 @@ def exception_handler(response: requests.models.Response):
                 else:
                     err_msg = error_entry.get('message', '')
             except ValueError:
-                err_msg = '{}'.format(response.text)
+                err_msg = f'{response.text}'
 
         raise DemistoException(err_msg)
 
+    def search_events(self, limit: int, start_date: str = None, end_date: str = None, next_link: str = None) -> dict:
+        return super()._http_request(
+            method='GET',
+            url_suffix=URL_SUFFIX['BASE'] + next_link if next_link else URL_SUFFIX['EVENTS'],
+            params=None if next_link else {'limit': limit, 'startDate': start_date, 'end_date': end_date}
+        )
+
 
 ''' HELPER FUNCTIONS '''
 
 
+def run_fetch_mechanism(client: Client, fetch_limit: int, next_link: str, start_date: int, end_date: int):
```

Comment:
very nice refactor

Reply:
> very nice refactor

https://github.com/demisto/content/pull/36077#discussion_r1798100485

Reply:
> > very nice refactor
> 
> https://github.com/demisto/content/pull/36077#discussion_r1798100485

From bb75deacf559f10f317ef17a1a4ef426140b6d3b Mon Sep 17 00:00:00 2001
From: TalZich <tzichlinsky@paloaltonetworks.com>
Date: Mon, 1 Jul 2024 15:03:08 +0300
Subject: [PATCH 01/45] checkout

---
 .../AtlassianConfluenceCloud.py               | 51 ++++++++++++++++++-
 1 file changed, 49 insertions(+), 2 deletions(-)

diff --git a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
index 8be42c06bcc9..5e4e37ab6da3 100644
--- a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
+++ b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
@@ -28,8 +28,10 @@
     "CONTENT": "/wiki/rest/api/content",
     "USER": "/wiki/rest/api/search/user?cql=type=user",
     "SPACE": "/wiki/rest/api/space",
-    "PRIVATE_SPACE": "/wiki/rest/api/space/_private"
+    "PRIVATE_SPACE": "/wiki/rest/api/space/_private",
+    "EVENTS": "/wiki/rest/api/audit/"
 }
+
 MESSAGES = {
     "REQUIRED_URL_FIELD": "Site Name can not be empty.",
     "NO_RECORDS_FOUND": "No {} were found for the given argument(s).",
@@ -88,6 +90,9 @@
     "permanent delete": "trashed",
     "permanent delete draft": "draft"
 }
+VENDOR = "Atlassian"
+PRODUCT = "Confluence"
+
 ''' CLIENT CLASS '''
 
 
@@ -143,6 +148,9 @@ def exception_handler(response: requests.models.Response):
 
         raise DemistoException(err_msg)
 
+    def search_events(self, limit: int, start: int = None, start_time: str = None) -> List[Dict]:
+        return self.http_request(urlsuffix=URL_SUFFIX['EVENTS'])
+
 
 ''' HELPER FUNCTIONS '''
 
@@ -1333,6 +1341,30 @@ def confluence_cloud_group_list_command(client: Client, args: Dict[str, str]) ->
         raw_response=response_json)
 
 
+def fetch_events(client: Client, last_run: dict[str, Any], limit: int) -> tuple[Dict, List[Dict]]:
+    last_index = last_run.get('last_index')
+
+    if last_index:
+        start = last_index + 1
+        events = client.search_events(start=start, limit=limit)
+
+    else:
+        start_time = str(round((time.time() - 60) * 1000))
+        events = client.search_events(start_time=start_time, limit=limit)
+
+    last_index += len(events)
+    demisto.debug(f'Fetched event with id: {last_index}.')
+
+    # Save the next_run as a dict with the last_fetch key to be stored
+    next_run = {'last_index': last_index}
+    demisto.debug(f'Setting next run {next_run}.')
+    return next_run, events
+
+
+def confluence_cloud_get_events():
+    pass
+
+
 ''' MAIN FUNCTION '''
 
 
@@ -1379,7 +1411,7 @@ def main() -> None:
             'confluence-cloud-space-list': confluence_cloud_space_list_command,
             'confluence-cloud-comment-create': confluence_cloud_comment_create_command,
             'confluence-cloud-content-create': confluence_cloud_content_create_command,
-            'confluence-cloud-space-create': confluence_cloud_space_create_command
+            'confluence-cloud-space-create': confluence_cloud_space_create_command,
         }
         command = demisto.command()
         args = demisto.args()
@@ -1391,6 +1423,21 @@ def main() -> None:
             return_results(test_module(client))
         elif command in commands:
             return_results(commands[command](client, args))
+
+        elif command == 'fetch-events':
+            last_run = demisto.getLastRun()
+            next_run, events = fetch_events()
+
+        elif command == 'confluence-cloud-get-events':
+            should_push_events = argToBoolean(args.get('should_push_events'))
+            events, results = confluence_cloud_get_events()
+            return_results(results)
+            if should_push_events:
+                send_events_to_xsiam(
+                    events,
+                    vendor=VENDOR,
+                    product=PRODUCT
+                )
     # Log exceptions and return errors
     except Exception as e:
         return_error(f'Failed to execute {demisto.command()} command.\nError:\n{str(e)}')

From 29f72a76061c0f864b76cbce21f1bd8001c9fcc5 Mon Sep 17 00:00:00 2001
From: TalZich <tzichlinsky@paloaltonetworks.com>
Date: Wed, 3 Jul 2024 16:55:14 +0300
Subject: [PATCH 02/45] Add unit tests, yet to finish

---
 .../AtlassianConfluenceCloud.py               |   58 +-
 .../AtlassianConfluenceCloud_test.py          |   38 +-
 .../test_data/collector/api_responses.json    | 1010 +++++++++++++++++
 3 files changed, 1094 insertions(+), 12 deletions(-)
 create mode 100644 Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/test_data/collector/api_responses.json

diff --git a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
index 5e4e37ab6da3..d0ee1f580b91 100644
--- a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
+++ b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud.py
@@ -148,13 +148,31 @@ def exception_handler(response: requests.models.Response):
 
         raise DemistoException(err_msg)
 
-    def search_events(self, limit: int, start: int = None, start_time: str = None) -> List[Dict]:
-        return self.http_request(urlsuffix=URL_SUFFIX['EVENTS'])
+    def search_events(self, limit: int, start_index: int = None, start_date: str = None) -> Dict:
+        return super()._http_request(
+            method='GET',
+            urlsuffix=URL_SUFFIX['EVENTS'],
+            params={'limit': limit, 'start': start_index, 'startDate': start_date}
+        )
 
 
 ''' HELPER FUNCTIONS '''
 
 
+def add_time_to_events(events: List[Dict] | None):
+    """
+    Adds the _time key to the events.
+    Args:
+        events: List[Dict] - list of events to add the _time key to.
+    Returns:
+        list: The events with the _time key.
+    """
+    if events:
+        for event in events:
+            create_time = arg_to_datetime(arg=event.get('creationDate'))
+            event['_time'] = create_time.strftime(DATE_FORMAT) if create_time else None
+
+
 def get_error_message(errors):
     err_msg = ""
     for error in errors:
@@ -1345,12 +1363,12 @@ def fetch_events(client: Client, last_run: dict[str, Any], limit: int) -> tuple[
     last_index = last_run.get('last_index')
 
     if last_index:
-        start = last_index + 1
-        events = client.search_events(start=start, limit=limit)
+        start_index = last_index + 1
+        events = client.search_events(start_index=start_index, limit=limit).get('results')
 
     else:
-        start_time = str(round((time.time() - 60) * 1000))
-        events = client.search_events(start_time=start_time, limit=limit)
+        start_date = str(round((time.time() - 60) * 1000))
+        events = client.search_events(start_date=start_date, limit=limit).get('results')
 
     last_index += len(events)
     demisto.debug(f'Fetched event with id: {last_index}.')
@@ -1361,8 +1379,23 @@ def fetch_events(client: Client, last_run: dict[str, Any], limit: int) -> tuple[
     return next_run, events
 
 
-def confluence_cloud_get_events():
-    pass
+def get_events(client: Client, args: dict) -> tuple[list[Dict], CommandResults]:
+    limit = args.get('limit', 50)
+    start_index = args.get('start')
+    start_date = args.get('start_date')
+    if start_index and start_date:
+        raise ValueError('Please provide either start_index or start_date, not both.')
+    if not start_index and not start_date:
+        events = client.search_events(limit=limit, start_index=0).get('results')
+        return events, CommandResults(readable_output=tableToMarkdown('Events', events, removeNull=True))
+
+    if start_index:
+        events = client.search_events(limit=limit, start_index=start_index).get('results')
+        return events, CommandResults(readable_output=tableToMarkdown('Events', events, removeNull=True))
+
+    if start_date:
+        events = client.search_events(limit=limit, start_date=start_date).get('results')
+        return events, CommandResults(readable_output=tableToMarkdown('Events', events, removeNull=True))
 
 
 ''' MAIN FUNCTION '''
@@ -1417,7 +1450,7 @@ def main() -> None:
         args = demisto.args()
         strip_args(args)
         remove_nulls_from_dictionary(args)
-
+        max_events_per_fetch = params.get('max_events_per_fetch', 10000)
         if command == 'test-module':
             # This is the call made when pressing the integration Test button.
             return_results(test_module(client))
@@ -1426,11 +1459,14 @@ def main() -> None:
 
         elif command == 'fetch-events':
             last_run = demisto.getLastRun()
-            next_run, events = fetch_events()
+            next_run, events = fetch_events(client, last_run, max_events_per_fetch)
+            add_time_to_events(events)
+            send_events_to_xsiam(events, vendor=VENDOR, product=PRODUCT)
+            demisto.setLastRun(next_run)
 
         elif command == 'confluence-cloud-get-events':
             should_push_events = argToBoolean(args.get('should_push_events'))
-            events, results = confluence_cloud_get_events()
+            events, results = get_events(client, args)
             return_results(results)
             if should_push_events:
                 send_events_to_xsiam(
diff --git a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud_test.py b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud_test.py
index 3b925c016a6d..31858574ecdb 100644
--- a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud_test.py
+++ b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/AtlassianConfluenceCloud_test.py
@@ -1,8 +1,10 @@
 import io
 
 import pytest
+from pytest_mock import MockerFixture
+
 from CommonServerPython import *
-from AtlassianConfluenceCloud import Client, URL_SUFFIX, MESSAGES
+from AtlassianConfluenceCloud import Client, URL_SUFFIX, MESSAGES, fetch_events, get_events
 from test_data import input_data
 
 BASE_URL = "https://dummy.atlassian.com"
@@ -975,3 +977,37 @@ def test_confluence_cloud_content_list_command_when_when_object_not_present(requ
     assert response.outputs_key_field == "id"
     assert response.outputs == expected_context_output
     assert response.readable_output == expected_readable_output
+
+
+def test_fetch_events_with_last_index(mocker: MockerFixture):
+    last_run = {'last_index': 10}
+    limit = 5
+
+    expected_start_index = last_run['last_index'] + 1
+
+    client.search_events.return_value = {'results': expected_events}
+
+    next_run, events = fetch_events(client, last_run, limit)
+
+    assert next_run == {'last_index': expected_start_index + len(expected_events)}
+    assert events == expected_events
+
+    client.search_events.assert_called_once_with(start_index=expected_start_index, limit=limit)
+
+
+def test_fetch_events_without_last_index():
+    client = MagicMock()
+    last_run = {}
+    limit = 5
+
+    expected_start_date = str(round((time.time() - 60) * 1000))
+    expected_events = [{'id': 1}, {'id': 2}]
+
+    client.search_events.return_value = {'results': expected_events}
+
+    next_run, events = fetch_events(client, last_run, limit)
+
+    assert next_run == {'last_index': len(expected_events)}
+    assert events == expected_events
+
+    client.search_events.assert_called_once_with(start_date=expected_start_date, limit=limit)
diff --git a/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/test_data/collector/api_responses.json b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/test_data/collector/api_responses.json
new file mode 100644
index 000000000000..11aaa6e1fb67
--- /dev/null
+++ b/Packs/AtlassianConfluenceCloud/Integrations/AtlassianConfluenceCloud/test_data/collector/api_responses.json
@@ -0,0 +1,1010 @@
+{
+    "get-audit-records": {
+        "results": [
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+                    "accountId": "<string>",
+                    "accountType": "<string>",
+                    "externalCollaborator": true,
+                    "isExternalCollaborator": true,
+                    "publicName": "<string>"
+                },
+                "remoteAddress": "<string>",
+                "creationDate": 59,
+                "summary": "<string>",
+                "description": "<string>",
+                "category": "<string>",
+                "sysAdmin": true,
+                "superAdmin": true,
+                "affectedObject": {
+                    "name": "<string>",
+                    "objectType": "<string>"
+                },
+                "changedValues": [
+                    {
+                        "name": "<string>",
+                        "oldValue": "<string>",
+                        "hiddenOldValue": "<string>",
+                        "newValue": "<string>",
+                        "hiddenNewValue": "<string>"
+                    }
+                ],
+                "associatedObjects": [
+                    {
+                        "name": "<string>",
+                        "objectType": "<string>"
+                    }
+                ]
+            },
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+                    "accountId": "<string>",
+                    "accountType": "<string>",
+                    "externalCollaborator": true,
+                    "isExternalCollaborator": true,
+                    "publicName": "<string>"
+                },
+                "remoteAddress": "<string>",
+                "creationDate": 59,
+                "summary": "<string>",
+                "description": "<string>",
+                "category": "<string>",
+                "sysAdmin": true,
+                "superAdmin": true,
+                "affectedObject": {
+                    "name": "<string>",
+                    "objectType": "<string>"
+                },
+                "changedValues": [
+                    {
+                        "name": "<string>",
+                        "oldValue": "<string>",
+                        "hiddenOldValue": "<string>",
+                        "newValue": "<string>",
+                        "hiddenNewValue": "<string>"
+                    }
+                ],
+                "associatedObjects": [
+                    {
+                        "name": "<string>",
+                        "objectType": "<string>"
+                    }
+                ]
+            },
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+                    "accountId": "<string>",
+                    "accountType": "<string>",
+                    "externalCollaborator": true,
+                    "isExternalCollaborator": true,
+                    "publicName": "<string>"
+                },
+                "remoteAddress": "<string>",
+                "creationDate": 59,
+                "summary": "<string>",
+                "description": "<string>",
+                "category": "<string>",
+                "sysAdmin": true,
+                "superAdmin": true,
+                "affectedObject": {
+                    "name": "<string>",
+                    "objectType": "<string>"
+                },
+                "changedValues": [
+                    {
+                        "name": "<string>",
+                        "oldValue": "<string>",
+                        "hiddenOldValue": "<string>",
+                        "newValue": "<string>",
+                        "hiddenNewValue": "<string>"
+                    }
+                ],
+                "associatedObjects": [
+                    {
+                        "name": "<string>",
+                        "objectType": "<string>"
+                    }
+                ]
+            },
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+                    "accountId": "<string>",
+                    "accountType": "<string>",
+                    "externalCollaborator": true,
+                    "isExternalCollaborator": true,
+                    "publicName": "<string>"
+                },
+                "remoteAddress": "<string>",
+                "creationDate": 59,
+                "summary": "<string>",
+                "description": "<string>",
+                "category": "<string>",
+                "sysAdmin": true,
+                "superAdmin": true,
+                "affectedObject": {
+                    "name": "<string>",
+                    "objectType": "<string>"
+                },
+                "changedValues": [
+                    {
+                        "name": "<string>",
+                        "oldValue": "<string>",
+                        "hiddenOldValue": "<string>",
+                        "newValue": "<string>",
+                        "hiddenNewValue": "<string>"
+                    }
+                ],
+                "associatedObjects": [
+                    {
+                        "name": "<string>",
+                        "objectType": "<string>"
+                    }
+                ]
+            },
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+                    "accountId": "<string>",
+                    "accountType": "<string>",
+                    "externalCollaborator": true,
+                    "isExternalCollaborator": true,
+                    "publicName": "<string>"
+                },
+                "remoteAddress": "<string>",
+                "creationDate": 59,
+                "summary": "<string>",
+                "description": "<string>",
+                "category": "<string>",
+                "sysAdmin": true,
+                "superAdmin": true,
+                "affectedObject": {
+                    "name": "<string>",
+                    "objectType": "<string>"
+                },
+                "changedValues": [
+                    {
+                        "name": "<string>",
+                        "oldValue": "<string>",
+                        "hiddenOldValue": "<string>",
+                        "newValue": "<string>",
+                        "hiddenNewValue": "<string>"
+                    }
+                ],
+                "associatedObjects": [
+                    {
+                        "name": "<string>",
+                        "objectType": "<string>"
+                    }
+                ]
+            },
+            {
+                "author": {
+                    "type": "user",
+                    "displayName": "<string>",
+                    "operations": {},
+                    "username": "<string>",
+                    "userKey": "<string>",
+